MaxText/layers/attentions.py

```diff
@@ -284,12 +417,12 @@ def apply_attention(
       query = query.astype(jnp.float32)
       key = key.astype(jnp.float32)
 
-    # QK Product, a.k.a `attn_weights`: [batch, num_heads, q_length, kv_length]
+    # QK Product, a.k.a `attn_weights`: [batch, num_kv_heads, num_heads_per_group, q_length, kv_length]
```

Comment:
nit: there would be cases that batch size reshaped [b, ..., kv_heads, group, ...]

Also I think you mean num_heads_per_group -> number of grouped query. as num_heads = num_kv_heads * num_groups.

Reply:
I'm calling it `num_query_heads_per_kv_head` now, I hope that helps!

I'm not sure I follow your nit but your nit scares me! Is there a where reshape misfires?

Reply:
not sure what's wrong with num_groups, the paper https://arxiv.org/pdf/2305.13245.pdf calls it's groups.